kafka/src/main/java/org/axonframework/extensions/kafka/eventhandling/tokenstore/TokenStoreState.java

```diff
@@ -0,0 +1,333 @@
+     * @param consumerConfiguration used to configure the Kafka consumer
+     * @param producerConfiguration used to configure the Kafka producer
+     */
+    TokenStoreState(Executor executor,
+                    String topic,
+                    TemporalAmount claimTimeout,
+                    Map<String, Object> consumerConfiguration,
+                    Map<String, Object> producerConfiguration,
+                    Duration writeTimeOut
+    ) {
+        this.executor = executor;
+        this.topic = topic;
+        this.claimTimeout = claimTimeout;
+        this.consumerConfiguration = consumerConfiguration;
+        this.producerConfiguration = producerConfiguration;
+        this.writeTimeOutMillis = writeTimeOut.toMillis();
+    }
+
+    Future<Boolean> send(TokenUpdate update) {
+        String key = String.format(KEY_FORMAT, update.getProcessorName(), update.getSegment());
+        CompletableFuture<Boolean> future = new CompletableFuture<>();
+        writeResult.put(update.getId(), future);
+        try {
+            producer.updateAndGet(p -> {
+                if (p == null) {
+                    return new KafkaProducer<>(producerConfiguration);
+                }
+                return p;
+            }).send(new ProducerRecord<>(topic, key, update)).get(writeTimeOutMillis, TimeUnit.MILLISECONDS);
+        } catch (InterruptedException e) {
+            writeResult.remove(update.getId());
+            future.complete(false);
+            Thread.currentThread().interrupt();
+        } catch (ExecutionException | TimeoutException e) {
+            logger.warn("error sending token update '{}[{}]' to Kafka",
+                        update.getProcessorName(),
+                        update.getSegment(),
+                        e);
+            writeResult.remove(update.getId());
+            future.complete(false);
+        }
+        return future;
+    }
+
+    Optional<TokenUpdate> getCurrent(String processorName, int segment) {
+        waitTillReady();
+        return Optional.ofNullable(state.get(processorName)).map(s -> s.get(segment));
+    }
+
+    int[] fetchSegments(String processorName) {
+        waitTillReady();
+        return Optional.ofNullable(state.get(processorName)).map(this::toPrimitiveIntArray).orElse(new int[0]);
+    }
+
+    Collection<TokenUpdate> fetchAll(String processorName) {
+        waitTillReady();
+        return Optional.ofNullable(state.get(processorName)).map(Map::values).orElse(Collections.emptyList());
+    }
+
+    void start() {
+        if (isRunning.compareAndSet(false, true)) {
+            executor.execute(this::startConsumer);
+            producer.set(new KafkaProducer<>(producerConfiguration));
+        }
+    }
+
+    void close() {
+        isRunning.set(false);
+        producer.updateAndGet(p -> {
+            if (p != null) {
+                p.close();
+            }
+            return null;
+        });
```

Comment:
Shouldn't we shutdown the `Executor` here too? If the default `Executor` is used, of course. If the user configures their own we should leave that responsibility to them.

Reply:
Added a `shutdownAction`, which defaults to shutting it down.